data/blog/2025/2025-07-23-benefits-of-middleware-over-mvc.md

```diff
@@ -0,0 +1,132 @@
+They can **process a request in a predefined sequence**.
+This is useful for **implementing complex processing logic**, like data processing, logging and authentication in a clean and manageable way.
+
+## Asynchronous Processing
+
+Middleware architectures can help you handle asynchronous processing, which is nowadays increasingly important for **building high-performance web applications**.
+A good example for this scenario is when you need to manage many concurrent connections, while allowing other parts of the application to run in the meantime.
+The downside is that this adds complexity and overhead in the code logic and error handling, making it harder to debug and test, but it can provide benefits in specific cases.
+
+## Better Suited for API-First Development
+
+**MVC is tightly coupled** to the web/MVC context which **difficult to reuse or test** your logic in CLI scripts, workers or other applications.
+**Middleware** can often be a **better fit for API-first development**, where the main interaction with the application is through APIs rather than traditional web pages.
+This is the recommended approach for handling different types of requests, such as RESTful APIs, GraphQL and WebSocket connections.
+
+## Framework Lock-in
+
+MVC has multiple **components tied to the framework**:
+
+- Request & response classes.
+- Routing system.
+- Controllers or service containers.
+- Lifecycle.
+
+Conversely, **middleware is reusable** across currently active frameworks with little or no changes thanks to:
+
+- **Implementing standards** like PSR-7 and PSR-11 which also helps **integrate with future PHP frameworks** that implement them.
+- **Decoupling of handlers** which means they **can be used by API, CLI, worker processes or unit tests** without modification and while not being dependent on HTTP requests, routes, frameworks or controllers.
+- **View rendering is optional or external**, since middleware cares about HTTP requests and responses, not about how they are rendered.
+- The **middleware pipeline**, like the one based on PSR-15, **improves flexibility, customizability and testability** for execution flows.
+- **Easier unit testing** without bootstrapping the framework thanks to middleware being small, self-contained, and decoupled from the full framework environment.
+
+## Implementing migrations
+
+One of the most beneficial advantages of middleware versatility is that it enables you to **migrate your codebase progressively**.
+We all know that migrations can be costly and labor-intensive, but middleware allows you to perform the migration one component at a time, **with little to no disruption** to the overall platform functionality.
+Basically, the old system is kept in place, while the components are replaced with new implementations using middleware.
+
+This approach enables breaking down large tasks into smaller, more manageable and easily testable pieces.
+It's called the **strangle strategy** (or strangler pattern).
```

Comment:
It's actually the [Strangler Fig](https://martinfowler.com/bliki/StranglerFigApplication.html) pattern.